### DISCUSSION TITLE
AWS assumed role not authorised to perform cognito-idp:AdminCreateUser

### DISCUSSION THREAD

**Discussion by alakdam07** [serverless/serverless#10866] [Q&A] [answered]
I am using [serverless-framework](https://www.serverless.com/) for my app. I have one  PreSignUp lambda function. When user will create an account, after press the submit button that lambda function will trigger. 

In my `serverless.yaml` file IAM role policy I attached the `cognito-idp:AdminCreateUser`.

When ever I press submit button, I got this error: `Unrecognizable lambda output`. In my cloudwatch shows me this error: 

> "User:
> arn:aws:sts::...
> is not authorized to perform: cognito-idp:AdminCreateUser on resource:
> arn:aws:cognito-idp:.....
> because no identity-based policy allows the
> cognito-idp:AdminCreateUser action"

I don't understand what I am doing wrong.

**This is my lambda function**

    import AWS from 'aws-sdk';
    import joi from 'joi';

    const Cognito = new AWS.CognitoIdentityServiceProvider();

    const PreSignUp = async (event: any, context: any) => {
      // NOTE: Cognito expects Username to be the user's email

      /**
       * @param This process.env.COGNITO coming from serverless.yaml file. line: 18
       */
      const userPoolId = process.env.COGNITO ? process.env.COGNITO : '';

      const UNSAFE_EMAIL = event.request.userAttributes.email;
      const test = event.request.userAttributes;

      console.log({ UNSAFE_EMAIL, userPoolId, event, test });

      // Normalize email and language
      const UNSAFE_TRIMMED_EMAIL = UNSAFE_EMAIL.trim();

      const schema = joi.object().keys({
        email: joi.string().trim().email({ minDomainSegments: 2 }).required(),
      });

      const validationResult = schema.validate({
        email: UNSAFE_TRIMMED_EMAIL,
      });

      if (validationResult.error) {
        console.log(JSON.stringify(validationResult.error, null, 2));
      }

      console.log('finish the validation error');

      // Validation successful, change variable names to reflect
      const VALIDATED_EMAIL = UNSAFE_TRIMMED_EMAIL;
      // const VALIDATED_LANGUAGE = UNSAFE_TRIMMED_LANGUAGE;

      // Cognito params
      // Username is the user's email
      // email is also required in UserAttributes in order to send confirmation
      // DesiredDeliveryMediums is required to send confirmation
      const params = {
        UserPoolId: userPoolId,
        Username: VALIDATED_EMAIL,
        UserAttributes: [
          {
            Name: 'email',
            Value: VALIDATED_EMAIL,
          },
        ],
        DesiredDeliveryMediums: ['EMAIL'],
        ClientMetadata: {
          role: 'customer',
        },
      };

      // Attempt to create user in Cognito
      try {
        const authRes = await Cognito.adminCreateUser(params).promise();
        console.log('Success: ', JSON.stringify(authRes, null, 2));
        // callback(null, event);
        context.done(null, event);
      } catch (err) {
        console.log('Error: ', JSON.stringify(err, null, 2));
        return {
          statusCode: 400,
          body: JSON.stringify({
            error: true,
            error_message: err.message,
          }),
        };
      }
    };

    export default PreSignUp;

**This is my YAML file**

    service: serverless-graphql-book

    package:
      individually: true
      include:
        - src/**

    provider:
      name: aws
      profile: ${env:profile}
      runtime: nodejs14.x
      stage: ${env:stage}
      region: eu-north-1
      timeout: 30
      apiName: ${self:service.name}-${self:provider.stage}
      environment:
        ITEM_TABLE: ${self:service}-items-${self:provider.stage}
        COGNITO: ${self:service.name}-pool
      iamRoleStatements:
        - Effect: Allow
          Action:
            - dynamodb:Query
            - dynamodb:PutItem
            - dynamodb:UpdateItem
            - dynamodb:DeleteItem
            - lambda:InvokeFunction
            - cognito-idp:CreateUserPoolClient
            - cognito-idp:DeleteUserPoolClient
            - cognito-idp:AdminConfirmSignUp
            - cognito-idp:AdminInitiateAuth
            - cognito-idp:AdminDeleteUser
            - cognito-idp:AdminCreateUser ## HERE IS MY POLICY
            - cognito-idp:AdminSetUserPassword
          Resource: 'arn:aws:dynamodb:${opt:region, self:provider.region}:*:table/${self:provider.environment.ITEM_TABLE}'
        - Effect: Allow
          Action:
            - dynamodb:Query
          Resource: 'arn:aws:dynamodb:${opt:region, self:provider.region}:*:table/${self:provider.environment.ITEM_TABLE}/index/*'

      apiGateway:
        shouldStartNameWithService: true

    functions:

      ## Create user
      PreSignUp: ## THIS IS THE LAMBDA FUNCTION 
        handler: src/handlers/PreSignUp.default
        events:
          - cognitoUserPool:
              pool: ${self:provider.environment.COGNITO}
              trigger: PreSignUp
              existing: true
        ## post Confirmation
      PostConfirmation:
        handler: src/handlers/postConfirmation.default
        events:
          - cognitoUserPool:
              pool: ${self:provider.environment.COGNITO}
              trigger: PostConfirmation
              existing: true

    resources:
      Resources:
        CognitoUserPool: ## THIS IS MY USER POOL
          Type: AWS::Cognito::UserPool
          Properties:
            MfaConfiguration: OFF ## Multi factor Authentocation
            UserPoolName: ${self:provider.environment.COGNITO}
            AutoVerifiedAttributes:
              - email
            UsernameAttributes:
              - email
            Policies:
              PasswordPolicy:
                MinimumLength: 6
                RequireLowercase: False
                RequireNumbers: True
                RequireSymbols: False
                RequireUppercase: True
        CognitoUserPoolClient:
          Type: AWS::Cognito::UserPoolClient
          Properties:
            ClientName: ${self:provider.environment.COGNITO}
            UserPoolId:
              Ref: CognitoUserPool
            ExplicitAuthFlows:
              - ADMIN_NO_SRP_AUTH
            GenerateSecret: false

    plugins:
      - serverless-webpack
      - serverless-offline
      - serverless-plugin-warmup
      - serverless-dotenv-plugin
      - serverless-prune-plugin

**Comment by jeremycare** [comment, marked as answer]
HI @alakdam07,

On your IAM policy, you only specify the DynamoDB resource. You need to add the Cognito identity pool in the Ressource field too!

**Reply by alakdam07** [reply]
Thank you @jeremycare,

I figured it out yesterday. Forget to closed it. Thank you again spending time on my issue

**Reply by alakdam07** [reply]
It suddenly stop working :D, I don't get it why,

```
- Effect: Allow
      Action:
        - cognito-idp:CreateUserPoolClient
        - cognito-idp:DeleteUserPoolClient
        - cognito-idp:AdminConfirmSignUp
        - cognito-idp:AdminInitiateAuth
        - cognito-idp:AdminDeleteUser
        - cognito-idp:AdminCreateUser
        - cognito-idp:AdminSetUserPassword
      Resource: arn:aws:cognito-idp:eu-north-1:****:userpool/eu-north-1_****
```

Getting this error: ` because no identity-based policy allows the cognito-idp:AdminCreateUser action`

**Reply by jeremycare** [reply]
I don't think it's good practices to build arns lice that. I recommend you use serverless to get the arn:

```
            -  Fn::GetAtt: [ CognitoUserPool, Arn ]
```

**Reply by alakdam07** [reply]
Where should I add this line?
